okio/src/appleMain/kotlin/okio/Buffer.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import kotlin.math.min
+import kotlinx.cinterop.ByteVarOf
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.plus
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.usePinned
+import platform.Foundation.NSData
+import platform.Foundation.NSMutableData
+import platform.Foundation.create
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.memcpy
+
+fun Buffer.write(
+  source: NSData,
+  offset: Int = 0,
+  byteCount: Int = source.length.toInt()
+): Buffer = apply {
+  var offset = offset
+  checkOffsetAndCount(source.length.toLong(), offset.toLong(), byteCount.toLong())
+
+  val limit = offset + byteCount
+  while (offset < limit) {
+    val tail = writableSegment(1)
+
+    val toCopy = minOf(limit - offset, Segment.SIZE - tail.limit)
+    tail.data.usePinned { pinned ->
+      memcpy(
+        __dst = pinned.addressOf(tail.limit),
+        __src = (source.bytes as CPointer<ByteVarOf<*>>) + offset,
```

Comment:
Also, it worth notice, that [`NSData.enumerateBytes(_:)`](https://developer.apple.com/documentation/foundation/nsdata/1408400-enumeratebytes) should be used instead of `NSData.bytes` in order to make it more memory efficient. Since the NSData can be backed by multiple not-contiguous buffers under-the-hood, and when you access such NSData using `bytes` property, all buffers will be flattened into one (creates new single contiguous buffer).

Reply:
Good call! Want to submit a PR?